### PR TITLE
Run and upload primer over packages

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Ref. https://github.com/PyCQA/pylint/issues/5364.

Next step for the primer.

Test run on local `main`:
https://github.com/DanielNoord/pylint/actions/runs/2396609324
Test run on a PR:
https://github.com/DanielNoord/pylint/actions/runs/2396610764

Basically, the PR run will now result in three files: the PR number, the output of the primer on the PR and the output of the `main` run against which the PR is being compared. 

This can be downloaded by a last and final step which takes all these files. Compares them and then comments on the relevant PR.

### CODE REVIEW

File: .github/workflows/primer_run_pr.yaml [hunk 1]
```diff
@@ -95,10 +100,24 @@ jobs:
...
+            const [matchArtifactMainOutput] = artifacts_main.data.artifacts.filter((artifact) =>
+              artifact.name == "primer_output");
+            const downloadWorkflowTwo = await github.rest.actions.downloadArtifact({
+               owner: context.repo.owner,
+               repo: context.repo.repo,
+               artifact_id: matchArtifactMainOutput.id,
+               archive_format: "zip",
+            });
+            fs.writeFileSync("primer_output_main.zip", Buffer.from(downloadWorkflowTwo.data));
+            return lastRunMain.head_sha
```
Comment (jacobtylerwalls) [inline .github/workflows/primer_run_pr.yaml:114]:
```suggestion
            return lastRunMain.head_sha;
```

File: tests/primer/primer_tool.py [hunk 2]
```diff
@@ -88,6 +100,35 @@ def _handle_prepare_command(self) -> None:
...
+        # We want to test all the code we can
+        enables = ["--enable-all-extensions", "--enable=all"]
+        # Duplicate code takes too long and is relatively safe
+        disables = ["--disable=duplicate-code"]
+        arguments = data.directories + data.pylint_args + enables + disables
+        if data.pylintrc_relpath:
+            arguments += [f"--rcfile={data.pylintrc_relpath}"]
+        output = StringIO()
+        reporter = JSONReporter(output)
+        Run(arguments, reporter=reporter, do_exit=False)
```
Comment (jacobtylerwalls) [inline tests/primer/primer_tool.py:129]: We don't want to lint every primer package twice, so I think we will need to turn off the python test that checks for crashes, yeah?

File: tests/primer/primer_tool.py [hunk 3]
```diff
@@ -88,6 +100,35 @@ def _handle_prepare_command(self) -> None:
...
+
+        packages: dict[str, list[dict[str, str | int]]] = {}
+
+        for package, data in self.packages.items():
+            output = self._lint_package(data)
+            packages[package] = output
+            print(f"Successfully primed {package}.")
+
+        with open(
+            PRIMER_DIRECTORY / f"output_{self.config.type}.txt", "a", encoding="utf-8"
```
Comment (jacobtylerwalls) [inline tests/primer/primer_tool.py:115]: If multiple PRs are each running the primer step, does this mean they will append their results to each other?

File: tests/primer/primer_tool.py (hunk 2 above)
Comment (DanielNoord) [reply tests/primer/primer_tool.py:129]: Yes! Although this still needs some work before we get there. We don't actually prime on `astroid` and `black` doesn't take too long. So for now I think this is fine.

File: tests/primer/primer_tool.py (hunk 3 above)
Comment (DanielNoord) [reply tests/primer/primer_tool.py:115]: No, they don't share their environments.

File: tests/primer/primer_tool.py [hunk 4]
```diff
@@ -88,6 +100,35 @@ def _handle_prepare_command(self) -> None:
             ) as f:
                 f.write(commit_string)
 
+    def _handle_run_command(self) -> None:
+        if Path(PRIMER_DIRECTORY / f"output_{self.config.type}.txt").exists():
+            os.remove(PRIMER_DIRECTORY / f"output_{self.config.type}.txt")
```
Comment (jacobtylerwalls) [inline tests/primer/primer_tool.py:105]: I guess this is what confused me -- if this isn't executing on the CI runners, is this for local testing? If so, should we just remove this and do "w" mode below?

**Comment by jacobtylerwalls** [comment]
Should we bump `setup-node` to v3? See:
```
Warning: Input 'version' has been deprecated with message: The version property will not be supported after October 1, 2019. Use node-version instead
```

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM,, but I think @jacobtylerwalls is right that we should try to remove the package from the normal primer if they are in the false positive primer.

**Review by jacobtylerwalls — APPROVED** [review]
I think it's unlikely we'll forget to turn off the double-primer runs once we enable running this over the full set of primer repositories, so if you wish to merge now and then sort that out in a future PR, 👍🏻
